x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriod.java

```diff
@@ -0,0 +1,156 @@
+        Map.entry(DATE_PERIOD, (fieldEval, source) -> fieldEval),
+        Map.entry(KEYWORD, (fieldEval, source) -> fieldEval),
+        Map.entry(TEXT, (fieldEval, source) -> fieldEval)
+    );
+
+    private static final String INVALID_INTERVAL_ERROR = "Invalid interval value in [{}], expected integer followed by one of "
+        + DATE_PERIODS
+        + " but got [{}]";
+
+    @FunctionInfo(
+        returnType = "date_period",
+        description = "Converts a string into a `date_period` value.",
+        examples = @Example(file = "convert", tag = "castToDatePeriod")
+    )
+    public ToDatePeriod(
+        Source source,
+        @Param(name = "string", type = { "date_period", "keyword", "text" }, description = "A string.") Expression v
+    ) {
+        super(source, v);
+    }
+
+    private ToDatePeriod(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected Map<DataType, BuildFactory> factories() {
+        return EVALUATORS;
+    }
+
+    @Override
+    protected final TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        TypeResolution typeResolution = isType(
+            field(),
+            factories()::containsKey,
+            sourceText(),
+            FIRST,
+            supportedTypesNames(supportedTypes())
+        );
+        if (field().foldable()) {
+            if (isString(field().dataType()) && isValidInterval(field().fold().toString()) == false) {
+                typeResolution = new TypeResolution(LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), field().fold()));
+            }
+
+        }
+        return typeResolution;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DATE_PERIOD;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new ToDatePeriod(source(), newChildren.get(0));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, ToDatePeriod::new, field());
+    }
+
+    @Override
+    public final Object fold() {
+        if (field() instanceof Literal l) {
+            Object v = l.value();
+            if (v instanceof BytesRef b) {
+                if (isValidInterval(b.utf8ToString()) == false) {
+                    throw new IllegalArgumentException(
+                        LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), b.utf8ToString())
+                    );
+                }
+                return EsqlDataTypeConverter.parseTemporalAmount(b.utf8ToString(), DATE_PERIOD);
+            }
+        } else if (field() instanceof ToDatePeriod && field().foldable()) {
+            return field().fold();
+        }
```

Comment:
nit: that's a bit more special casing than we normally have. (It's not wrong, though!)

But: we seem to allow `to_dateperiod(to_dateperiod(...))`. To be consistent, I think we should either allow general foldable expressions, or literals only. E.g., why wouldn't I be able to write something like `to_dateperiod(concat(3::string, " weeks"))` (use case: I want to parameterize the number of weeks in my query)

Reply:
Foldable expressions(like `Concat`) can be allowed here.